File: server/match_manager.lua
local Class  = require "common.extra_libs.hump.class"
local Server = require "server"
local log    = require "common.extra_libs.log"
local Match  = require "match"

local MatchManager = {}

-- Map: Client -> Match
local match_map = {}

function MatchManager.init()
    local function assertMatch(client)
        if match_map[client] == nil then
            log.warn('Client not on any match.')
            Server.kick(client)
        end
        return match_map[client] ~= nil
    end

    local actions_locked_schema = { 'array', 'string' }
    Server.on('actions locked', function(data, client)
        if not Server.checkSchema(client, actions_locked_schema, data) or not assertMatch(client) then return end
        match_map[client]:actionsLocked(client, data)
    end)

    local action_input_schema = { 'number', 'number' }
    Server.on('action input', function(data, client)
        if not Server.checkSchema(client, action_input_schema, data) or not assertMatch(client) then return end
        match_map[client]:actionInput(client, data)
    end)
end

function MatchManager.startMatch(cl_list)
    local m = Match(cl_list)
    for _, cl in ipairs(cl_list) do
        match_map[cl.client] = m
    end
end

return MatchManager

File: server/match_manager.ts
import * as Server from "./server";
import { Client } from "common/extra_libs/sock";

class Match {
    cl_list: Client[];
    actions: string[][];
    lock_count: number;

    constructor(c1: Client, c2: Client) {
        this.cl_list = [c1, c2];
        this.cl_list.forEach((c, i) => c.send('start game', i))
        this.actions = [];
        this.lock_count = 0;
    }

    actionLocked(i: number, actions: string[]) {
        this.lock_count++;
        this.actions[i] = actions;
        if (this.lock_count == this.cl_list.length) {
            this.cl_list.forEach(c => c.send('turn ready', this.actions));
            this.lock_count = 0;
            this.actions = [];
        }
    }

    actionInput(data: unknown, client: Client) {
        this.cl_list.forEach(cl => {
            if (cl !== client)
                cl.send('action input', data);
        })
    }
}

const match_map = new Map<Client, Match>();

export function init() {
    Server.on('actions locked', (data: { i: number, actions: string[] }, client: Client) => {
        match_map.get(client)!.actionLocked(data.i, data.actions);
    })

    Server.on('action input', (data, client: Client) => {
        match_map.get(client)!.actionInput(data, client);
    })
}

export function startMatch(c1: Client, c2: Client) {
    const m = new Match(c1, c2);
    match_map.set(c1, m);
    match_map.set(c2, m);
}